src/video/renderers/opengl3/gl3_renderer.c

```diff
@@ -238,28 +239,34 @@ static void capture_screenshot(void *userdata) {
     omf_free(buffer);
 }
 
+#define ASPECT_X (4.0f / 3.0f)
+#define ASPECT_Y (3.0f / 4.0f)
+
 /**
  * Set the viewport, and do screen-shakes here.
  */
 static inline void set_screen_viewport(const gl3_context *ctx) {
-    float move_ratio = ctx->screen_w / NATIVE_W;
     float native_aspect = ctx->screen_w / ctx->screen_h;
     int vp_x = 0;
     int vp_y = 0;
     int vp_w = ctx->viewport_w;
     int vp_h = ctx->viewport_h;
 
+    // aspect == 0 means 4:3
+    // aspect == 1 means stretch to window
     if(ctx->aspect == 0) {
-        // Set 4:3 aspect ratio, as requested
         if(native_aspect >= 1.0f) {
-            vp_w = ctx->viewport_h * 4 / 3;
-            vp_x = (ctx->viewport_w - vp_w) / 2;
+            vp_w = min2(ctx->screen_w, ctx->viewport_h * ASPECT_X);
+            vp_h = vp_w * ASPECT_Y;
         } else {
-            vp_h = ctx->viewport_w * 3 / 4;
-            vp_y = (vp_h - ctx->viewport_h) / 2;
+            vp_h = min2(ctx->screen_h, ctx->viewport_w * ASPECT_Y);
+            vp_w = vp_h * ASPECT_X;
         }
```

Comment:
```suggestion
        vp_w = min2(ctx->screen_w, ctx->viewport_h * ASPECT_X);
        vp_h = min2(ctx->screen_h, ctx->viewport_w * ASPECT_Y);
```
to simplify

Reply:
First suggestion fails to produce 4:3 aspect ratios since sizes are not linked. Second suggestion probably works.

Reply:
Did some cleanups, and split this into a unit-tested helper function.